htsget-search/src/htsget/bam_search.rs

```diff
@@ -32,17 +32,36 @@ impl VirtualPositionExt for VirtualPosition {
   /// But when we need to translate it into a byte range, we need to make sure
   /// the reads falling inside that block are also included, which requires to know
   /// where that block ends, which is not trivial nor possible for the last block.
-  /// The simple solution goes through adding the maximum BGZF block size,
-  /// so we don't loose any read (although adding extra unneeded reads to the query results).
-  fn bytes_range_end(&self) -> u64 {
-    self.compressed() + Self::MAX_BLOCK_SIZE
+  /// The solution used here goes through reading the records starting at that point.
+  /// That way we can know where the blocks ends and only add to the query result
+  /// the needed blocks. If for some reason we can't read correctly the
+  /// records we fall back to adding the maximum BGZF blocks size
+  fn bytes_range_end(&self, reader: &mut bam::Reader<File>) -> u64 {
+    get_next_block_position(self.clone(), reader)
+      .unwrap_or(self.compressed() + Self::MAX_BLOCK_SIZE)
   }
 
   fn to_string(&self) -> String {
     format!("{}/{}", self.compressed(), self.uncompressed())
   }
 }
 
+fn get_next_block_position(
+  block_position: VirtualPosition,
+  reader: &mut bam::Reader<File>,
+) -> Option<u64> {
+  reader.seek(block_position).ok()?;
+  let next_block_index = loop {
+    let mut record = bam::Record::default();
+    reader.read_record(&mut record).ok()?;
```

Comment:
How come getting positions needs mutable readers? Would it be possible to have it immutable?

Reply:
Ah, yeah, I see. Cool then, no worries nor changes needed here :)